libs/state_estimator/src/state_estimator.cpp

```diff
@@ -269,6 +281,160 @@ namespace STATE_ESTIMATOR {
+        auto xVariance = std::accumulate(result.xList.begin(), result.xList.begin() + result.xSize, 0.0f,
+                                        [xMean](float acc, float x) { return acc + std::pow(x - xMean, 2); }) / result.xSize;
+        auto yVariance = std::accumulate(result.yList.begin(), result.yList.begin() + result.ySize, 0.0f,
+                                        [yMean](float acc, float y) { return acc + std::pow(y - yMean, 2); }) / result.ySize;
+
+        float totalVariance = xVariance + yVariance;
+
+        return {totalVariance, xMean, yMean};
+    }
+
+    Pose StateEstimator::filterPositions(Pose odometryEstimate, Pose localisationEstimate){
+        /**
+         * Combines the odometry and localization estimates to produce a filtered position estimate.
+         * This method uses a weighted average approach to merge the estimates, potentially improving
+         * the accuracy of the robot's perceived position within the arena.
+         *
+         * @param odometryEstimate A Pose struct representing the position estimate based on odometry.
+         * @param localisationEstimate A Pose struct representing the position estimate based on localization.
+         * @return A Pose struct representing the filtered position and heading of the robot.
+         */
+
+        Pose filteredPosition;
+        filteredPosition.x = (1 - localisation_weighting) * odometryEstimate.x + localisation_weighting * localisationEstimate.x;
+        filteredPosition.y = (1 - localisation_weighting) * odometryEstimate.y + localisation_weighting * localisationEstimate.y;
+        filteredPosition.heading = odometryEstimate.heading;
+        return filteredPosition;
+    }
+
+
+    StateEstimator::PermutationResult StateEstimator::createPermutation(
+            int permutation,
+            const std::array<float, NUM_TOF_SENSORS>& xPositions,
+            const std::array<float, NUM_TOF_SENSORS>& yPositions) {
+        /**
+         * Generates a specific permutation of valid X and Y position estimates based on the given permutation number.
+         * The permutation number is interpreted as a binary representation, where each bit indicates whether a position
+         * from the input X or Y positions list is selected. Positions corresponding to '1' bits are added to the Y list,
+         * and positions corresponding to '0' bits are added to the X list. The function returns arrays containing the
+         * selected positions along with the actual count of valid positions in each array.
+         *
+        * @param permutation The permutation number, interpreted as a binary pattern for selecting X or Y positions.
+        * @param xPositions The array of potential X positions.
+        * @param yPositions The array of potential Y positions.
+        * @return A structure containing the X and Y lists along with their counts of valid elements.
+        */
+        std::array<float, NUM_TOF_SENSORS> xList, yList;
```

Comment:
minor: `xList` and `yList` are uninitialsed here - that's probably fine, as long as `std::array<float, NUM_TOF_SENSORS>` has more members than `xSize` (which it almost certainly does). To be defensive, it's a good idea to initialise the contents of the `xList` and `yList` arrays:
```suggestion
std::array<float, NUM_TOF_SENSORS> xList = {0.0f};
std::array<float, NUM_TOF_SENSORS> yList = {0.0f};
```